File: 1290.convert-binary-number-in-a-linked-list-to-integer.py
#
# @lc app=leetcode id=1290 lang=python3
#
# [1290] Convert Binary Number in a Linked List to Integer
#

# @lc code=start
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def getDecimalValue(self, head: ListNode) -> int:
        # a= ''
        # while head:
        #     a += str(head.val)
        #     head = head.next
        # return int(a, 2)

        #todo alternative soltion
        #it is just too slow but it works just fine mua
        cnt = 0
        ans = 0
        binary_list = []
        while head:
            binary_list.append(int(head.val))
            cnt += 1
            head = head.next
        
        for ele in binary_list:
            ans += ele * (2 ** (cnt-1))
            cnt -= 1
        return ans
        
# @lc code=end

File: 1290.convert-binary-number-in-a-linked-list-to-integer.java
/*
 * @lc app=leetcode id=1290 lang=java
 *
 * [1290] Convert Binary Number in a Linked List to Integer
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    //算法技巧在于数学运算
    public int getDecimalValue(ListNode head) {
        // head = reverse(head);
        // int ans = 0;
        // int base = 1;
        // while (head != null) {
        //     ans += head.val * base;
        //     base *= 2;
        //     head = head.next;
        // }
        // return ans;
        int ans = 0;
        while(head != null){
            ans = ans*2 + head.val;
            head = head.next;
        }
        return ans;
    }
    // private ListNode reverse(ListNode head){
    //     ListNode prev = null;
    //     ListNode cur = head;
    //     while(cur != null){
    //         ListNode next = cur.next;
    //         cur.next = prev;
    //         prev = cur;
    //         cur = next;
    //     }
    //     head = prev;
    //     return head;
    // }
}
// @lc code=end

